dlib/media/ffmpeg_abstract.h

```diff
@@ -0,0 +1,955 @@
+            float                                   sample_rate{0};
+            std::chrono::system_clock::time_point   timestamp{};
+        };
+
+        // ---------------------------------------------------------------------------------------------------
+
+        struct codec_details
+        {
+            /*!
+                WHAT THIS OBJECT REPRESENTS
+                    This object informs on available codecs provided by the installation of ffmpeg dlib is linked against.
+            !*/
+            std::string codec_name;
+            bool supports_encoding;
+            bool supports_decoding;
+        };
+
+        struct device_details
+        {
+            /*!
+                WHAT THIS OBJECT REPRESENTS
+                    This object informs on available devices provided by the installation of ffmpeg dlib is linked against.
+            !*/
+
+            struct instance
+            {
+                std::string name;
+                std::string description;
+            };
+
+            std::string device_type;
+            std::vector<instance> devices;
+        };
+        
+        std::vector<std::string> list_protocols();
+        /*!
+            ensures
+                - returns a list of all registered ffmpeg protocols
+        !*/
+
+        std::vector<std::string> list_demuxers();
+        /*!
+            ensures
+                - returns a list of all registered ffmpeg demuxers
+        !*/
+
+        std::vector<std::string> list_muxers();
+        /*!
+            ensures
+                - returns a list of all registered ffmpeg muxers
+        !*/
+        
+        std::vector<codec_details> list_codecs();
+        /*!
+            ensures
+                - returns a list of all registered ffmpeg codecs with information on whether decoding and/or encoding is supported.
+                  Note that not all codecs support encoding, unless your installation of ffmpeg is built with third party library
+                  dependencies like libx264, libx265, etc.
+        !*/
+
+        std::vector<device_details> list_input_devices();
+        /*!
+            ensures
+                - returns a list of all registered ffmpeg input devices and available instances of those devices
+        !*/
+
+        std::vector<device_details> list_output_devices();
+        /*!
+            ensures
+                - returns a list of all registered ffmpeg output devices and available instances of those devices
+        !*/
+
+        // ---------------------------------------------------------------------------------------------------
+
+        template <class ImageContainer>
+        void convert(const frame& f, ImageContainer& image)
+        /*!
+            requires
+                - f.is_image() == true
+                - f.pixfmt() == pix_traits<typename image_traits<ImageContainer>::pixel_type>::fmt
+
+            ensures
+                - converts a frame object into array2d<rgb_pixel>
+        !*/
```

Comment:
```suggestion
        template <class image_type>
        void convert(const frame& f, image_type& image)
        /*!
            requires
                - f.is_image() == true
                - f.pixfmt() == pix_traits<typename image_traits<image_type>::pixel_type>::fmt
                - image_type == an image object that implements the interface defined in
                  dlib/image_processing/generic_image.h 
            ensures
                - Copies the image content from f into the provided image.
        !*/
```
Something like that.  One of the precondition is the type contract of the image.  And this thing supports any of them, not just array2d.  And might as well call the type `image_type` since that's what the rest of the library uses for this kind of argument.

Reply:
Done